### PR TITLE
Removed communication and geometry globals from domain decomposition

### PR DESCRIPTION
**Pull request by fweik**
If I wouldn't know better I'd say that this code was specifically crafted to be hard to refactor.

Description of changes:
- Removed geometry globals from dd
- Removed communication globals from dd
- Modified dd so that the periodicity does not change the communication pattern
- Treat periodicity change as geometry change
- Moved `get_mi_*` stuff to `BoxGeometry`

### CODE REVIEW

File: src/core/event.cpp [hunk 1]
```diff
@@ -325,8 +325,7 @@ void on_parameter_change(int field) {
     break;
   case FIELD_NODEGRID:
     grid_changed_n_nodes();
-    cells_on_geometry_change(CELL_FLAG_GRIDCHANGED);
-    break;
+    cells_re_init(CELL_STRUCTURE_CURRENT, cell_structure.min_range);
   case FIELD_MINNUMCELLS:
   case FIELD_MAXNUMCELLS:
     cells_re_init(CELL_STRUCTURE_CURRENT, cell_structure.min_range);
```
Comment (jngrad) [inline src/core/event.cpp:327]: is this fallthrough intentional? `cells_re_init()` is now called twice with the same arguments if case `FIELD_NODEGRID` matches. Maybe you meant to delete L328?

File: src/utils/include/utils/mpi/cart_comm.hpp [hunk 2]
```diff
@@ -121,6 +121,38 @@ cart_neighbors(const boost::mpi::communicator &comm) {
 
   return ret;
 }
+
+/**
+ * @brief Information about a cartesian communicator.
+ *
+ * Memebers coorespond to the output arguments of
```
Comment (jngrad) [inline src/utils/include/utils/mpi/cart_comm.hpp]:
```suggestion
 * Members correspond to the output arguments of
```

**Review by jngrad — CHANGES_REQUESTED** [review]
There are several parts of the ghost communication code where I'll trust you :)

File: src/core/event.cpp (hunk 1 above)
Comment (fweik) [reply src/core/event.cpp:327]: Ah yes, good catch.

**Comment by fweik** [comment]
It's clear to me that this is a little bit of a leap of faith, but it has to be done to get that under control. We can then subsequently find out what the actual pre- and postconditions are and add tests. Also all of this has fairly good coverage by the integration tests.

**Review by jngrad — APPROVED** [review]

**Pull request merged**